HPCC-18593 Create helper library to digitally sign text

RussWhitehead opened the pull request:
Create a general purpose library that can use optional public/private
key files to digitally sign text strings. This will initially be used
to authenticate Dali requests when passwords are no longer available
or change every 30 seconds

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

RussWhitehead committed:
HPCC-18593 Create helper library to digitally sign text

Create a general purpose library that can use optional public/private
key files to digitally sign text strings. This will initially be used
to authenticate Dali requests when passwords are no longer available
or change every 30 seconds

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

RussWhitehead commented:
@ghalliday @mckellyln  I was unable to resuscitate the previous branch so I was forced to create another.
For reference, the previous PR was https://github.com/hpcc-systems/HPCC-Platform/pull/10915
Please rereview

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 41d3393e588f448c69fa45ca71ad88a73a60c850
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |764 |764 |0 |
|test (thor) |678 |678 |0 |
|test (roxie) |791 |791 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  183 sec (00:03:03) |   55 sec (00:00:55) |    6 sec (00:00:06) |   38 sec (00:00:38) | 1129 sec (00:18:49) |   32 sec (00:00:32) | 1458 sec (00:24:18) |

mckellyln commented on system/security/digisign/digisign.cpp:
```diff
@@ -0,0 +1,253 @@
+
+        EVP_PKEY* pKey = EVP_PKEY_new();
+        if (nullptr == pKey)
+        {
```
Could we leak rsa here ?

mckellyln commented on system/security/digisign/digisign.cpp:
```diff
@@ -0,0 +1,253 @@
+#ifdef _USE_OPENSSL
+        EVP_MD_CTX * RSACtx;
+        EVP_PKEY * EVPKey;
+        digiInit(true, privateKeyBuff.str(), passphraseBuff.str(), &RSACtx, &EVPKey);
```
Just a note, passphraseBuff from file may be encrypted and it is used here as is.  Perhaps a comment that passphraseBuff may need to be decrypted here before being used.

mckellyln commented on system/security/digisign/digisign.cpp:
```diff
@@ -0,0 +1,253 @@
+#ifdef _USE_OPENSSL
+        EVP_MD_CTX * RSACtx;
+        EVP_PKEY * EVPKey;
+        digiInit(false, publicKeyBuff.str(), passphraseBuff.str(), &RSACtx, &EVPKey);
```
same comment here

mckellyln commented:
@RussWhitehead one question and comment.
This works for when specifiying the public key by way of the cert file, correct ?
And then you will add addtional public key pem file format option in another PR, correct ?

RussWhitehead commented:
@mckellyln Correct, I will update the call to queryHPCCPKIKeyFiles with the PR (HPCC-19212)  that adds the public key file to the environment.conf. Please rereview

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a2e96b2f9feb3744f75fe373680053f972e68fe7
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |765 |765 |0 |
|test (thor) |679 |679 |0 |
|test (roxie) |792 |792 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   32 sec (00:00:32) |  182 sec (00:03:02) |   55 sec (00:00:55) |    6 sec (00:00:06) |   37 sec (00:00:37) | 1136 sec (00:18:56) |   33 sec (00:00:33) | 1481 sec (00:24:41) |

mckellyln commented on system/security/digisign/digisign.cpp:
```diff
@@ -0,0 +1,260 @@
+        const char * cert=0, * privKey=0, * passPhrase=0;
+        bool rc = queryHPCCPKIKeyFiles(&cert, &privKey, &passPhrase);
+        if (!isEmptyString(cert))
+            publicKeyFile.set(cert);
```
Just a question - queryHPCCPKIKeyFiles() looks for "HPCCCertFile" keyword.
Is this a "cert" file format or a "public key" file format ?

RussWhitehead replied:
Public key, in PEM format

mckellyln replied:
ok, then should the conf file keyword be changed from "HPCCCertFile" to something more appropriate, like "HPCCPubKeyFile" ?

RussWhitehead replied:
Will change queryHPCCPKIKeyFiles() via separate Jira

ghalliday commented on system/security/digisign/digisign.cpp:
```diff
@@ -0,0 +1,260 @@
+        verifyingConfigured = false;
+#ifdef _USE_OPENSSL
+        //query private key file location from environment.conf
+        const char * cert=0, * privKey=0, * passPhrase=0;
```
Much cleaner to assign nullptr, instead of 0.

RussWhitehead replied:
For now, the signature is created when a user is authenticated, and signature verification occurs in Dali when a file request is made from one of the engines.

mckellyln commented on system/security/digisign/digisign.cpp:
```diff
@@ -0,0 +1,260 @@
+        if (!isEmptyString(privKey))
+            privateKeyFile.set(privKey);
+        if (!isEmptyString(passPhrase))
+            passphraseBuffEnc.set(passPhrase);//MD5 encrypted passphrase
```
I think I was wrong before - isn't the passPhrase returning from queryHPCCPKIKeyFiles() already decrypted ?

RussWhitehead replied:
It is already decrypted by queryHPCCPKIKeyFiles.  I think it would be better if queryHPCCPKIKeyFiles returned it encrypted, and we decrypt it as needed.  Thoughts?

RussWhitehead replied:
I will change queryHPCCPKIKeyFiles to return the encrypted password via separate Jira

mckellyln commented on system/security/digisign/digisign.cpp:
```diff
@@ -0,0 +1,260 @@
+        {
+            StringBuffer ppDec;
+            if (!isEmptyString(passphraseEnc))
+                decrypt(ppDec, passphraseEnc);
```
See comment above about this.
```
jlib_decl bool queryHPCCPKIKeyFiles(const char * *  _certificate,//HPCCCertFile
                                    const char * *  _privateKey, //HPCCPrivateKeyFile
                                    const char * *  _passPhrase) //HPCCPassPhrase
{
    const IProperties & conf = queryEnvironmentConf();
    if (_certificate)
        *_certificate = conf.queryProp("HPCCCertFile");
    if (_privateKey)
        *_privateKey = conf.queryProp("HPCCPrivateKeyFile");;
    if (_passPhrase)
    {
        CriticalBlock b(HPCCpassPhraseCrit);
        if (HPCCpassPhrase.isEmpty())
        {
            const char *passPhrasePtr = conf.queryProp("HPCCPassPhrase");
            if (!isEmptyString(passPhrasePtr))
                decrypt(HPCCpassPhrase, passPhrasePtr);
        }
        *_passPhrase = HPCCpassPhrase.str();
    }
    return true;
}
```

ghalliday reviewed (commented):
One style comment.

Otherwise the code looks good - assuming that it is not designed to be used very frequently.  If the design does require it to be used frequently, then almost all the code in digiInit should be being done once, cached in the objects and reused in each of the calls.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 5b500ddec5d66437f8e94ab9a70d271749adadef
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |765 |765 |0 |
|test (thor) |679 |679 |0 |
|test (roxie) |792 |792 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   18 sec (00:00:18) |  177 sec (00:02:57) |   56 sec (00:00:56) |    6 sec (00:00:06) |   39 sec (00:00:39) | 1125 sec (00:18:45) |   33 sec (00:00:33) | 1454 sec (00:24:14) |

RussWhitehead commented:
@mckellyln @ghalliday Please rereview

ghalliday commented:
@RussWhitehead can you confirm how frequently you expect this to be called?
Otherwise looks good to me.  @mckellyln

ghalliday commented:
It should have a unit test, but happy to merge if a jira is created to add one later.

RussWhitehead commented:
@ghalliday There is no external interface for this, so not sure how to add a unit test. It will be used internally to authenticate users within ESP and when an engine calls Dali.  Any ideas?

mckellyln commented:
I am ok with code also, but note:
1). PR for HPCCPKIKeyFiles() to use Public Key still needs to support same functionality as querySecuritySettings() call it could replace - but that is a separate PR.
2). As Gavin mentions - is this a performance hit to Dali ?  Should much of the functionality be done once and the decodedSig be saved so that the potentially often called compare be as fast as possible ?
ie just this -
```
     int match = EVP_DigestVerifyFinal(RSACtx, (unsigned char *)decodedSig.str(), decodedSig.length());
```

RussWhitehead commented:
@mckellyln 
1) Opened https://track.hpccsystems.com/browse/HPCC-19298 for addingcertificate
2) Opened https://track.hpccsystems.com/browse/HPCC-19297 for performance concerns

RussWhitehead commented:
@mckellyln @ghalliday Are we okay to merge this PR (see above Jira's I opened)

mckellyln commented:
Yes, please link those issues back to this one.
I am ok to mege.

RussWhitehead commented:
@richardkchapman @ghalliday Please merge

ghalliday commented:
@RussWhitehead can you confirm how frequently you expect this to be called?

RussWhitehead commented:
@ghalliday Signature will be created when a user is authenticated, and will be verified by Dali on any file request that includes creds (if Dali authentication is enabled)

ghalliday commented:
@RussWhitehead 
That sounds like a situation where performance may be very significant.

Because this is a general purpose helper library I don't think it can be merged until it includes some unit tests and investigates (and likely improves) the performance.

As far as adding unit tests goes they should use the CPPUNIT framework (e.g., see examples in jlibtests.cpp).  To make it easier I would restructure the code.  I would modify CDigitalSignatureManager so that it takes cert, privKey and passPhrase as parameters to the constructor (the actual contents, rather than the filenames), and provide function(s) to create instances.  That way you can then make a self contained test with the keys provided inline.  It would probably require that you switch to creating the global object in MODULE_INIT, rather than a static constructor.

Once you have unit tests you can also sensibly test how long the operations take.  Given the API has been provided to reuse the objects I suspect there is a significant benefit to be had from reusing it.  You may find it easier to have two implementations of the interface - one if OPEN_SSL is enabled, and a dummy implementation if it is not.

RussWhitehead commented:
@ghalliday @mckellyln I found that I have to create the context every time, otherwise when I try to encrypt or decrypt using any text/key combo other what what was first used, it fails.  However, I also added CPPUNIT tests, and found the following timings while creating the context every time
digiSign() 1000 iterations took 4402MS
digiverify 1000 iterations took 154MS
Considering I create a signature once, when the user is authenticated, and check it when received by Dali, these timings seem acceptable.

RussWhitehead commented:
@ghalliday @mckellyln I should also comment that when a user signature is verified by Dali, I will add it to the (LDAP) cache, and make sure to look it up and verify BEFORE invoking the digiVerify code

RussWhitehead commented:
@ghalliday @mckellyln See my comment above regarding performance.  I added CPPUNIT tests, and refactored to create/destroy the context with every call.  
Please rereview

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: c3914528386d91f7d555fe2c04df36a55064c2b7
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |84 |84 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |758 |758 |0 |
|test (thor) |685 |685 |0 |
|test (roxie) |798 |798 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  182 sec (00:03:02) |   55 sec (00:00:55) |    6 sec (00:00:06) |   38 sec (00:00:38) | 1183 sec (00:19:43) |   33 sec (00:00:33) | 1512 sec (00:25:12) |

ghalliday commented on system/security/digisign/digisign.cpp:
```diff
@@ -0,0 +1,273 @@
+#ifdef _USE_OPENSSL
+        //query private key file location from environment.conf
+        const char * pubKey, * privKey, * passPhrase;
+        bool rc = queryHPCCPKIKeyFiles(nullptr, &pubKey, &privKey, &passPhrase);
```
@RussWhitehead 
This doesn't compile when I try it locally.  It is defined as 
```
extern jlib_decl bool queryHPCCPKIKeyFiles(const char * *  _certificate,//HPCCCertFile
                                           const char * *  _privateKey, //HPCCPrivateKeyFile
                                           const char * *  _passPhrase);//HPCCPassPhrase
```
And I don't see a change in this PR to add an extra parameter.

RussWhitehead replied:
Try refreshing your master, the new argument was added via https://github.com/hpcc-systems/HPCC-Platform/pull/10952

ghalliday commented on system/security/digisign/digisign.cpp:
```diff
@@ -0,0 +1,273 @@
+        return verifyingConfigured;
+    }
+
+    bool digiInit(bool isSigning, const char * passphraseEnc, EVP_MD_CTX * * ctx, EVP_PKEY * * PKey)
```
I don't think this function will compile if openssl is not installed/available.

ghalliday commented:
@RussWhitehead the unit test needs to be self contained.  After deleting the extra parameter and running the unit test I had the error:

    Digisign IException thrown:digiSign:Creating Digital Signatures not configured

The tests need to be able to run even without access to the configuration.  (Even with access to the configuration the signing key should also be private, so should not be accessible from user code anyway.)  That is the reason for suggesting above the restructuring of the class to allow the information to be passed in for the test cases.

(It would also be a more generally useful class if they were passed in.)

RussWhitehead commented:
@ghalliday I made the following changes
1) Made digiInit() private, and wrapped in #ifdef USE_OPENSSL
2) Added two ctors and factory methods to create for given key files, or key buffers

In the unit test
1) Embedded a public and private key buffer in the cpp file
2) Construct the signer using the embedded key buffers
Please rereview

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6325045763dd8f73bd55f9c337da42843aa394a9
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |84 |84 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   31 sec (00:00:31) |  180 sec (00:03:00) |   55 sec (00:00:55) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1131 sec (00:18:51) |   16 sec (00:00:16) | 1440 sec (00:24:00) |

ghalliday commented:
@RussWhitehead that is good - I can now run the regression tests on my machine.

A couple of high level comments
a) I would only have a single constructor for the CDigitalSignatureManager class.  The versions that take it from files/environment can be implemented as factory functions.  The reason for suggesting this is the code that checks if keyBuffs are empty is the only part of the code that is not thread safe.  If you can guarantee that they are always loaded then your critical section can be removed.  (It can be for your unit tests since they don't use those functions).  That is very significant - more than 10x speed up on my machine.

You could have different constructors and load the files at construction, but then reporting errors is harder.  You could also have a separate function to load the files after construction before returning the pointer - but the class doesn't really need to know about the original files, so it would be cleaner to not pass them.
Note, there are some helpers in jmutex.hpp (Singleton) to make it easy to implement a thread safe singleton that uses a factory method. (See comments at the end of that file.)

b) Loading the key at construction and reusing it.
I added some timings to see what effect this would have.  For 1000 iterations on my machine:

Signing the initialization took 19ms, signing took 1.9s - initialization is 1% of the total, so negligable.
Verifying the initialization took 10.5ms, verifying took 12.9ms - key initialization is 80% of the total.

So although the verifying is much faster than the signing, about 80% of the time is spent initialising the key which could be created once and used for each call.  It is worth getting that overhead as low as possible by caching it in the constructor.

RussWhitehead commented:
@ghalliday I made the change you recommended, giving the CDigitalSignatureManager class a single ctor that accepts the key buffer contents, and added factories that accept a key file name, or key file buffers.
Regarding the reuse of the keys, see my comment from March 15th ("I found that I have to create the context every time, otherwise when I try to sign/verify using any text/key combo other what what was first used, it fails.")
Please rereview

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: b01967d2c0b525ec80775c4ad7ef54fc6abf35ff
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |84 |84 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   25 sec (00:00:25) |  182 sec (00:03:02) |   55 sec (00:00:55) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1134 sec (00:18:54) |   16 sec (00:00:16) | 1439 sec (00:23:59) |

ghalliday commented on system/security/digisign/digisign.cpp:
```diff
@@ -294,17 +235,58 @@ extern "C"
+        static Owned<IDigitalSignatureManager> dsm;
+
+#ifdef _USE_OPENSSL
+        if (nullptr == dsm.get())
```
This is not thread safe - despite appearances - and why I pointed you at the jlib classes!

Theroetically the dsm.get() within the critical section does not need to be re-evaluated - which means that two threads could create an instance, update dsm, causing the first thread to crash since it would be using a version of the class which had just been freed by the second thread.

ghalliday replied:
If you are going to code it yourself you need to use an atomic..

ghalliday commented:
@RussWhitehead yes you are right that the context cannot be shared.  But as far as I understand it the key can be loaded once and shared.  (That is what I added timings for in the figures above.)

RussWhitehead commented:
@ghalliday When I try to reuse the key, the code crashes on the second call to both signing and verifying. Can I have a look at what you did?  Basically I call everything in diginit up to and including EVP_PKEY_assign_RSA() once only

ghalliday commented:
I restructured your code, and I didn't have any problems getting it to work (other than ensuring I didn't call EVP_CLEANUP any more). (See my digSign branch for the code).

However there are problems with the multithreaded access. @mckellyln did you know this?  Because of this my code cores when testing that with a higher number of iterations. (My guess is it may be because of the add-ref implementation.)

See https://www.openssl.org/blog/blog/2017/02/21/threads/ for a blog post.  And https://www.openssl.org/docs/man1.0.2/crypto/threads.html for more documentation.

So I think the options are one of:
a) Create the keys on each call.
b) Protect each digiVerify/sign call with a critical section.
c) Implement the locking callbacks required by the library.

@mckellyln do you agree?

@RussWhitehead 
Because of the threading issues I suspect it may be worth going for (a) since it is the simplest, but with clear comments to indicate why.   I will add some other comments onto the PR.  (My branch also shows the use of the singleton class in the second commit).

ghalliday commented on system/security/digisign/digisign.cpp:
```diff
@@ -294,17 +235,58 @@ extern "C"
     {
-        static CDigitalSignatureManager digitalSignatureManager;
-        return & digitalSignatureManager;
+        static Owned<IDigitalSignatureManager> dsm;
```
static variables are generally better outside a function (they have unusual initialisation semantics if they are within a function.)

RussWhitehead commented on system/security/digisign/digisign.cpp:
```diff
@@ -294,17 +235,58 @@ extern "C"
+
+        if (!isEmptyString(_pubKey))
+        {
+            CriticalBlock b(DSMFilesCrit);
```
agreed

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8a2dec131476df5d8d4961d0ad4184b27379c82f
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |84 |84 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   16 sec (00:00:16) |  183 sec (00:03:03) |   56 sec (00:00:56) |    5 sec (00:00:05) |   21 sec (00:00:21) | 1139 sec (00:18:59) |   17 sec (00:00:17) | 1437 sec (00:23:57) |

RussWhitehead commented:
@ghalliday @mckellyln @richardkchapman  Can this be merged?

ghalliday commented on system/security/digisign/digisign.cpp:
```diff
@@ -230,70 +231,76 @@ class CDigitalSignatureManager : implements IDigitalSignatureManager, public CIn
-            }
-        }
+        std::call_once(dsmInitFlag, createDigitalSignatureManagerInstance, &dsm);
+	    return *dsm;
```
strange indentation

ghalliday commented on system/security/digisign/digisign.cpp:
```diff
@@ -230,70 +231,76 @@ class CDigitalSignatureManager : implements IDigitalSignatureManager, public CIn
+        std::call_once(dsmInitFlag, createDigitalSignatureManagerInstance, &dsm);
+	    return *dsm;
+#else
+        return nullptr;
```
this will not compile if _USE_OPENSSL is not defined

ghalliday reviewed (commented):
A couple of minor comments.

ghalliday commented:
@RussWhitehead a couple of minor comments.  Please can you rebase onto the current master (to remove commit that is already merged) and then squash the remaining commits.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: f95dd2b1ece700867fc89634bbd20fe09c9f998a
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |84 |84 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   30 sec (00:00:30) |  182 sec (00:03:02) |   55 sec (00:00:55) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1125 sec (00:18:45) |   17 sec (00:00:17) | 1437 sec (00:23:57) |
